GoogleForCreators/web-stories-wp
# Data Structure: reduce redundancy when storing fonts

## Context

This is a follow-up to https://github.com/GoogleForCreators/web-stories-wp/pull/12131.

Right now when adding 10 text elements all with the same font, there will be 10 font objects being stored, all with the same information. That's very redundant.

Instead of this, consider storing all existing fonts at the story level instead of on each element to reduce redundancy.



## Summary

Keeps as much of the existing code i.e.  calls to calcFontMetrics(element) the same while reducing how we store the font data.

### When Saving or Updating
- `getStoryPropsToSave.js`
    - Save fonts data at story level
    - Remove "extra" element font data   keep only font family 

👉  At this point we should have clean / reduced font storage

### When we re-load
- `useLoadStory.js`
    - Re-populate the element font data pulling fonts data from story level

👉  At this point we should have clean font data at the story level and the data back on the elements --- meaning code that wants to get the data from the element (font) can do so vs #12445 where I introduced a new elementFontData param for passing font data.

**See before and after:**
https://github.com/GoogleForCreators/web-stories-wp/pull/12625#issuecomment-1320961376



## Relevant Technical Choices



## To-do



## User-facing changes



## Testing Instructions






This PR can be tested by following these steps:

This PR makes updates to font strorage we shouldn't see any visual changes while editing, previewing or outputting a story.

1. Create / edit a story
2. Preview / insert a story 
3. Look for visual regressions
4. When saving check the network request see:

https://github.com/GoogleForCreators/web-stories-wp/pull/12625#issuecomment-1330540671

If you have access to the database you can check the way stories are saved i.e. 
https://github.com/GoogleForCreators/web-stories-wp/pull/12625#issuecomment-1320961376






## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #12261
